modules/gateway/ip.go

```diff
@@ -0,0 +1,97 @@
+package gateway
+
+import (
+	"net"
+	"time"
+
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/errors"
+)
+
+// discoverPeerIP is the handler for the discoverPeer RPC. It returns the
+// public ip of the caller back to the caller. This allows for peer-to-peer ip
+// discovery without centralized services.
+func (g *Gateway) discoverPeerIP(conn modules.PeerConn) error {
+	conn.SetDeadline(time.Now().Add(connStdDeadline))
+	host, _, err := net.SplitHostPort(conn.RemoteAddr().String())
+	if err != nil {
+		return errors.AddContext(err, "failed to split host from port")
+	}
+	return encoding.WriteObject(conn, host)
+}
+
+// managedIPFromPeers asks the peers the node is connected to for the node's public ip address. If not enough peers are available
+func (g *Gateway) managedIPFromPeers() (string, error) {
+	for {
+		// Check for shutdown signal.
+		select {
+		case <-g.peerTG.StopChan():
+			return "", errors.New("interrupted by shutdown")
+		default:
+		}
```

Comment:
Instead of being an infinite loop, this should be a loop that times out entirely after a few minutes. I'm thinking 5 minutes is probably a good amount of time.

Reply:
That makes be a bit nervous tbh. What if the gateway can't connect to peers for some reason and after 10 minutes it can? That means we stopped trying to discover our IP and the only way for the peer to get a valid ip is to restart the node.

Reply:
We have 2 issues then. The first is that means we may have a thread perpetually trying to collect an ip address, which for some reason isn't working. But the second is that we actually don't have any way right now of detecting if our ip address has changed. It should be uncommon, even for people with dynamic ip addresses (typically doesn't happen while you are online, but I believe it can if your ISP decides to do it). So we might want to have some loop that occasionally spins up another IP check.

Generally speaking, there is nothing in the function title `managedIPFromPeers` that indicates the function could potentially block for a long time (or even forever if something is wrong). Anything that has a super long return time should probably happen from a `threaded` function. For this use case, I think we can handle that pretty cleanly just by moving the outermost loop to the `threaded` function and leaving the rest of this logic here.